jwst/tweakreg/astrometric_utils.py

```diff
@@ -0,0 +1,179 @@
+import os
+import requests
+
+from astropy import table
+from astropy.table import Table
+from astropy.coordinates import SkyCoord
+from astropy import units as u
+
+from resample import resample_utils
+
+ASTROMETRIC_CAT_ENVVAR = "ASTROMETRIC_CATALOG_URL"
+DEF_CAT_URL = 'http://gsss.stsci.edu/webservices'
+
+if ASTROMETRIC_CAT_ENVVAR in os.environ:
+    SERVICELOCATION = os.environ[ASTROMETRIC_CAT_ENVVAR]
+else:
+    SERVICELOCATION = DEF_CAT_URL
+
+"""
+
+Primary function for creating an astrometric reference catalog.
+
+"""
+
+
+def create_astrometric_catalog(input_models, catalog="GAIADR2", output="ref_cat.ecsv",
+                               gaia_only=False, table_format="ascii.ecsv",
+                               existing_wcs=None, num_sources=None):
+    """Create an astrometric catalog that covers the inputs' field-of-view.
+
+    Parameters
+    ----------
+    input : str, list
+        Filenames of images to be aligned to astrometric catalog
+
+    catalog : str, optional
+        Name of catalog to extract astrometric positions for sources in the
+        input images' field-of-view. Default: GAIADR2. Options available are
+        documented on the catalog web page.
+
+    output : str, optional
+        Filename to give to the astrometric catalog read in from the master
+        catalog web service.  If None, no file will be written out.
+
+    gaia_only : bool, optional
+        Specify whether or not to only use sources from GAIA in output catalog
+
+    existing_wcs : `~stwcs.wcsutil.HSTWCS`
+        existing WCS object specified by the user
+
+    num_sources : int
+        Maximum number of brightest/faintest sources to return in catalog.
+        If `num_sources` is negative, return that number of the faintest
+        sources.  By default, all sources are returned.
+
+    Notes
+    -----
+    This function will point to astrometric catalog web service defined
+    through the use of the ASTROMETRIC_CATALOG_URL environment variable.
+
+    Returns
+    -------
+    ref_table : `~astropy.table.Table`
+        Astropy Table object of the catalog
+
+    """
+
+    # start by creating a composite field-of-view for all inputs
+    # This default output WCS will have the same plate-scale and orientation
+    # as the first chip in the list, which for WFPC2 data means the PC.
```

Comment:
This may need updating for JWST context. Probably something relevant to which, if any, NIRCam detector is normally first in the list. But if it's just the list of members as specified in the ASN table that's input to the `calwebb_image3` pipeline, the order can be semi-random.

Reply:
Done...